### PR TITLE
cephadm: add a new SMB service to the mgr module

### PR DESCRIPTION
**Pull request by phlogistonjohn**
<s>Depends on #54817</s>
Depends on #55719

Add initial support for deploying SMB gateways using Samba Container based daemons from a new smb service spec. An example service spec that I used for testing follows:

```yaml
service_type: smb
service_id: tango
placement:
  hosts: 
    - ceph0
spec:
  cluster_id: tango
  features:
    - domain
  config_uri: rados://.smb/tango/scc.toml
  custom_dns:
    - "192.168.76.204"
  join_sources:
    - "rados:mon-config-key:smb/config/tango/join1.json"
  include_ceph_users: 
    - client.smb1
```

The only feature flag supported is 'domain'. A deployment w/o that flag is a "users/groups" based deployment w/o active directory support. In the future a 'clustered' feature flag will added to support said feature. 

The content of the `config_uri` and `join_sources` fields are pseudo-URIs that point to objects stored in RADOS pool (`rados://` similar to how nfs is used on ceph) or through the rados api call to fetch data from the mon configuration key store (`rados:mon-config-key:`). In both cases the content of the objects is largely opaque to ceph orchestration and is defined by the [samba-container](https://github.com/samba-in-kubernetes/samba-container/) / [sambacc](https://github.com/samba-in-kubernetes/sambacc/) projects.

In the future I plan on submitting a new `smb` manager module that is functionally similar to the nfs module that will automate the construction of these configurations. You'll be able to define smb clusters, smb shares, etc through the `ceph` command line tool. Thus the intended layering works like so:

```
    Admin User
    (ceph commands)
     |
     |
     V
  +-----------------------+       +----------+
  | MGR: smb module  <----------->| RADOS    |
  |                   o   |       +----------+
  +-------------------|---+
  | MGR: orch module  V   |
  +-------------------|---+
  | MGR: cephadm mod  V   |
  +-------------------|---+
                      |
                      |
  +-------------------V---+
  |  NODE: cephadm        |
  |                 o     |
  +-----------------|-----+
  | NODE            V     |
  |  podman/docker        |
  + - - - - - - - - - - - +
  | NODE:                 |       +-------------+
  |           data <------------->| MDS/CephFS  |
  |                       |       +-------------+
  |  samba-container      |
  |                       |       +----------+
  |         config o------------->| RADOS    |
  +-----------------------+       +----------+
```

Each layer should largely stand on it's own so (for now) I'm planning on filing separate PRs for each item.  Because there is a logical dependency chain between each item we can review and discuss them independently but choose to merge only the final PR if convenient.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), <s>opened tracker ticket</s>
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
  - [x] Docs will be updated in a follow-up PR
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by rkachach** [comment]
jenkins test rook e2e

**Review by rkachach — DISMISSED** [review]
@phlogistonjohn please, can you have a look to the following error (rook e2e testing is failing):

`Error ENOENT: module 'rook' reports that it cannot run on the active manager daemon: cannot import name 'SMBSpec' (pass --force to force enablement)`

**Comment by phlogistonjohn** [comment]
> @phlogistonjohn please, can you have a look o the following error (rook e2e testing is failing):
> 
> `Error ENOENT: module 'rook' reports that it cannot run on the active manager daemon: cannot import name 'SMBSpec' (pass --force to force enablement)`

I looked at the code and I looked at the log from the jenkins job and I looked to see if there was an obvious place where I missed something (by looking for other spec types, like Nvemof). I do not see where the rook module would be trying to pull in the SMBSpec class, so it must be indirect. I will need a traceback/mgr logs to debug this further. How can I get that?

**Comment by phlogistonjohn** [comment]
^ @rkachach  any way I can get the mgr logs for these runs?

**Comment by rkachach** [comment]
> ^ @rkachach any way I can get the mgr logs for these runs?

Sorry @phlogistonjohn , I saw your message but I was doing some more investigation related to the error above. I think it's related to the rook e2e testing. I opened this PR to fix it https://github.com/ceph/ceph/pull/55393. It would be good to see if the patch from the PR fixes the error on your side.

**Comment by phlogistonjohn** [comment]
> > ^ @rkachach any way I can get the mgr logs for these runs?
> 
> Sorry @phlogistonjohn , I saw your message but I was doing some more investigation related to the error above. I think it's related to the rook e2e testing. I opened this PR to fix it #55393. It would be good to see if the patch from the PR fixes the error on your side.

Would it best if I cherry-pick your patch into this PR branch (for now)?

**Comment by phlogistonjohn** [comment]
> Would it be best if I cherry-pick your patch into this PR branch (for now)?

Since it's so easy to undo if it's a bad approach, that is what I'm going with for now.

**Comment by rkachach** [comment]
> > > ^ @rkachach any way I can get the mgr logs for these runs?
> > 
> > 
> > Sorry @phlogistonjohn , I saw your message but I was doing some more investigation related to the error above. I think it's related to the rook e2e testing. I opened this PR to fix it #55393. It would be good to see if the patch from the PR fixes the error on your side.
> 
> Would it be best if I cherry-pick your patch into this PR branch (for now)?

Sure :+1:, Thanks @phlogistonjohn.

**Comment by phlogistonjohn** [comment]
@rkachach `ceph rook orchestrator e2e tests` passed this time.

**Comment by rkachach** [comment]
> @rkachach `ceph rook orchestrator e2e tests` passed this time.

awesome, thanks @phlogistonjohn

**Comment by rkachach** [comment]
ceph rook orchestrator e2e tests

**Comment by rkachach** [comment]
jenkins test rook e2e

**Comment by phlogistonjohn** [comment]
After many failed attempts and much waiting I have a successful run of the new tests:

https://pulpito.ceph.com/phlogistonjohn-2024-03-05_18:46:11-orch:cephadm:smb-wip-phlogistonjohn-testing-2024-03-05-0803-distro-default-smithi/


Next steps are:
* to rebase & clean up the many temporary commits
* to add a test case that deploys a samba container without AD membership

**Comment by phlogistonjohn** [comment]
Another passing run with the cleaned up code & temporary commits squashed:
https://pulpito.ceph.com/phlogistonjohn-2024-03-06_12:45:19-orch:cephadm:smb-wip-phlogistonjohn-testing-2024-03-05-1601-distro-default-smithi/

I feel that the last task needed before I take the PR out of draft is to add a test case that deploys a samba container without AD membership (as noted above). So I think were very close.

In the meantime:

@rkachach - you still have a 'changes requested' on this PR. Can you please clear that? I'm assuming the request was specific to the failing rook test we already resolved.

@adk3798 I am fairly confident that the majority of the code in the cephadm binary and mgr module should be stable now. If anything turns up developing the final test I'll call it out specifically so that reviewers do not need to re-look at everything. I'm noting this just in case you have time to start looking before moving this out of draft. Thanks!

**Comment by phlogistonjohn** [comment]
https://pulpito.ceph.com/phlogistonjohn-2024-03-15_23:02:30-orch:cephadm:smb-wip-phlogistonjohn-testing-2024-03-15-1503-distro-default-smithi/

The errors are all log-scrape health warnings where a volume (temporarily?) lacks an application. I checked with @adk3798 and we consider this good enough to consider the deployment of smb containers to have succeeded. Finally moving this PR to reviewable.

File: doc/cephadm/services/smb.rst [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+SMB Service
+===========
+
+.. note:: Only the SMB3 protocol is supported.
+
+.. warning::
+
+    SMB support is under active development and many features may be
+    missing or immature. Additionally, a Manager module to automate
+    smb clusters and smb shares is in development. Once that feature
```
Comment (anthonyeleven) [inline doc/cephadm/services/smb.rst:13]: SMB uppercase?

File: doc/cephadm/services/smb.rst [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    A list of pre-defined terms enabling specific deployment characteristics.
+    An empty list is valid. Supported terms:
+
+    * ``domain``: Enable domain member mode
+
+config_uri
+    A string containing a (standard or de-facto) URI that identifies a
+    configuration source that should be loaded by the samba-container as the
+    primary configuration file.
+    Supported URI schemes include ``http:``, ``https:``, ``rados:``, and
```
Comment (anthonyeleven) [inline doc/cephadm/services/smb.rst:86]: Will this include the ability to specify the SMB protocol version?

File: doc/cephadm/services/smb.rst [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+join_sources
+    A list of strings with (standard or de-facto) URI values that will
+    be used to identify where authentication data that will be used to
+    perform domain joins are located. Each join source is tried in sequence
+    until one succeeds.
+    See ``config_uri`` for the supported list of URI schemes.
+
+custom_dns
+    A list of IP addresses that will be used as the dns servers for a Samba
```
Comment (anthonyeleven) [inline doc/cephadm/services/smb.rst:97]: DNS

File: doc/cephadm/services/smb.rst [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    container. This features allows the Samba Containers to integrate with
```
Comment (anthonyeleven) [inline doc/cephadm/services/smb.rst:98]: s/the//

File: doc/cephadm/services/smb.rst [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    Active Directory even if the ceph host nodes are not tied into the Active
```
Comment (anthonyeleven) [inline doc/cephadm/services/smb.rst:99]: Ceph

File: doc/cephadm/services/smb.rst [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    The cephx keys for each user in the list will automatically be added to
+    the keyring in the container.
+
+
+Configuring an SMB Service
+--------------------------
+
+.. warning::
+
+   A Manager module for smb is under active development. Once that module
```
Comment (anthonyeleven) [inline doc/cephadm/services/smb.rst:113]: SMB

File: doc/cephadm/services/smb.rst [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+* DNS is a critical component of Active Directory. If one is configuring the
+  smb service for domain membership, either the ceph host node must be
+  configured so that it can resolve the Active Directory (AD) domain or the
+  ``custom_dns`` option may be used. In both cases DNS hosts for the AD domain
+  must still be reachable from whatever network segment the ceph cluster is on.
+* Proper clustering/high-availability/"transparent state migration" is not yet
+  supported. If a placement causes more than service to be created these
+  services will act independently and may lead to unexpected behavior if clients
+  access the same files at once.
+* Services must bind to port 445. Running multiple smb services on the same node
```
Comment (anthonyeleven) [inline doc/cephadm/services/smb.rst:204]: SMB
Is this port UDP and/or TCP?

File: src/cephadm/tests/test_util_funcs.py [hunk 8]
```diff
@@ -906,3 +906,65 @@ def test_daemon_sub_id_systemd_names():
...
+
+
+@pytest.mark.parametrize(
+    "args,new_arg,expected",
+    [
+        (['--foo=77'], '--bar', ['--foo=77', '--bar']),
+        (['--foo=77'], '--foo=12', ['--foo=12']),
+        (
+            ['--foo=77', '--quux=later', '--range=2-5'],
+            '--quux=now',
```
Comment (anthonyeleven) [inline src/cephadm/tests/test_util_funcs.py:918]: Is that the great or the lesser Quux?

Comment (phlogistonjohn) [reply src/cephadm/tests/test_util_funcs.py:918]: The lesser northeastern gray quux of course!

File: doc/cephadm/services/smb.rst (hunk 7 above)
Comment (phlogistonjohn) [reply doc/cephadm/services/smb.rst:204]: Always TCP, I'll update.

File: src/cephadm/tests/test_util_funcs.py (hunk 8 above)
Comment (anthonyeleven) [reply src/cephadm/tests/test_util_funcs.py:918]: I figured this was an obscure Guy Steele reference

File: doc/cephadm/services/smb.rst (hunk 2 above)
Comment (phlogistonjohn) [reply doc/cephadm/services/smb.rst:86]: This config uri (and join sources) tell the samba container where to pull configuration properties from. It doesn't effect the what SMB protcol directly. One could add a custom config file enabling SMBv1. However, our plan is to only support SMBv2 & SMBv3 in the higher level management when it gets added.

File: doc/cephadm/services/smb.rst (hunk 3 above)
Comment (phlogistonjohn) [reply doc/cephadm/services/smb.rst]: done

File: doc/cephadm/services/smb.rst (hunk 4 above)
Comment (phlogistonjohn) [reply doc/cephadm/services/smb.rst]: done

File: doc/cephadm/services/smb.rst (hunk 5 above)
Comment (phlogistonjohn) [reply doc/cephadm/services/smb.rst]: done

File: doc/cephadm/services/smb.rst (hunk 6 above)
Comment (phlogistonjohn) [reply doc/cephadm/services/smb.rst]: ditto

File: src/cephadm/tests/test_util_funcs.py (hunk 8 above)
Comment (phlogistonjohn) [reply src/cephadm/tests/test_util_funcs.py:918]: Heh, I had to look that up. Apparently he's associated with that "metasyntactic variable" in particular but I wasn't aware of that off the top of my head. I thought adding lesser/greater to it made it sound like a bird species or something, thus my little joke. :-)

**Comment by phlogistonjohn** [comment]
jenkins test make check

**Comment by phlogistonjohn** [comment]
jenkins test api

**Review by adk3798 — COMMENTED** [review]
review of the `mgr/cephadm: fix test failure on newer python` commit onward as the cephadm binary stuff was reviewed in a prior PR. Review of the teuthology test is also quite limited by my lack of knowledge of SMB itself.

File: src/python-common/ceph/deployment/service_spec.py [hunk 9]
```diff
@@ -830,6 +832,7 @@ def _cls(cls: Type[ServiceSpecT], service_type: str) -> Type[ServiceSpecT]:
             'jaeger-collector': TracingSpec,
             'jaeger-query': TracingSpec,
             'jaeger-tracing': TracingSpec,
+            SMBSpec.service_type: SMBSpec,
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/python-common/ceph/deployment/service_spec.py:835]: In abstract, I always prefer to use a constant var / class var / enum over a plain string because then the language and static checkers can help find typos. Of course, like everyone does somtimes I get lazy ;-).

File: src/pybind/mgr/cephadm/services/smb.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class SMBService(CephService):
+    TYPE = 'smb'
+
+    def config(self, spec: ServiceSpec) -> None:
+        assert self.TYPE == spec.service_type
+        logger.warning('config is a no-op')
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/services/smb.py:22]: This is the only one I meant to leave at the warning level, because I couldn't figure out when this method was supposed to get called.
The others are just me being lazy and I forgot to turn them into debugs. Will fix.

File: src/pybind/mgr/cephadm/services/smb.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        self, smb_spec: SMBSpec, daemon_id: str, ceph_users: List[str]
+    ) -> Dict[str, str]:
+        ackc = self._allow_config_key_command(smb_spec.cluster_id)
+        wanted_caps = ['mon', f'allow r, {ackc}']
+        pools = list(self._pools_in_spec(smb_spec))
+        if pools:
+            wanted_caps.append('osd')
+            wanted_caps.append(
+                ', '.join(f'allow r pool={pool}' for pool in pools)
+            )
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/services/smb.py:114]: Sorry, could you rephrase the question, I'm finding it difficult to understand.

File: src/pybind/mgr/cephadm/module.py [hunk 12]
```diff
@@ -3256,7 +3258,8 @@ def _apply_service_spec(self, spec: ServiceSpec) -> str:
                 'elasticsearch': PlacementSpec(count=1),
                 'jaeger-agent': PlacementSpec(host_pattern='*'),
                 'jaeger-collector': PlacementSpec(count=1),
-                'jaeger-query': PlacementSpec(count=1)
+                'jaeger-query': PlacementSpec(count=1),
+                'smb': PlacementSpec(count=1),
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/module.py:3260]: See I hadn't even seen this comment when I wrote my other reply about getting lazy sometimes. :-)

File: src/pybind/mgr/orchestrator/module.py [hunk 13]
```diff
@@ -1820,6 +1821,46 @@ def _apply_jaeger(self,
...
+        format: Format = Format.plain,
+        no_overwrite: bool = False,
+        inbuf: Optional[str] = None,
+    ) -> HandleCommandResult:
+        """Apply an SMB network file system gateway service configuration."""
+
+        if not inbuf:
+            raise OrchestratorValidationError(
+                'unrecognized option -i; -h or --help for usage'
+            )
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/orchestrator/module.py:1844]: This is a copy-paste from other parts of the code. But I think what it's trying to do is block `-i` even though it's part of the command. If you pass `-i` infbuf is not-None. Thus the if block is true. If the block executes it raises an exception saying `-i` is no good (with misleading terms).

File: src/pybind/mgr/cephadm/services/smb.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/smb.py:114]: I'm curious about `if pools:` which implies having pools backing the SMB service is optional. I just wanted to confirm that's the intent, and that while you can specify pools for the SMB service, you don't want them to be required.

File: src/pybind/mgr/orchestrator/module.py [hunk 14]
```diff
@@ -1820,6 +1821,46 @@ def _apply_jaeger(self,
...
+            preview_only=dry_run,
+            cluster_id=cluster_id,
+            features=_features,
+            config_uri=config_uri,
+            join_sources=join_sources,
+            custom_dns=custom_dns,
+        )
+
+        spec.validate()  # force any validation exceptions to be caught correctly
+        # The previous comment makes no sense to JJM. But when in rome.
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/orchestrator/module.py:1860]: The `caught correctly` bit seems off/weird to me. Again, this was a copy-paste of other similar functions in this file, but I just couldn't leave that one unremarked. :-D

File: qa/suites/orch/cephadm/smb/tasks/deploy_smb_domain.yaml [hunk 15]
```diff
@@ -0,0 +1,88 @@
+roles:
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply qa/suites/orch/cephadm/smb/tasks/deploy_smb_domain.yaml:1]: OK. This is not the end of the teuthology work for sure. This is just the very start to get smb bootstrapped. We can think about how to better integrate (or not) later, IMO.

File: qa/suites/orch/cephadm/smb/tasks/deploy_smb_basic.yaml [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+roles:
+  # Test is for basic smb deployment & functionality. one node cluster is OK
+  - - host.a
+    - mon.a
+    - mgr.x
+    - osd.0
+    - osd.1
+    - client.0
+  # Reserve a host for acting as a domain controller
+  # TODO: ^ not really. Is there anything useful this host is doing?
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply qa/suites/orch/cephadm/smb/tasks/deploy_smb_basic.yaml:11]: Ah, I meant to remove the comment on the comment (and fix the original comment) after I realized we use the host purely as a client at the end of the test. But once the test "succeeded" I forgot all about that. Will fix.

File: src/pybind/mgr/orchestrator/module.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/orchestrator/module.py:1844]: Took a look in orchestrator/module.py and saw these two instances of `if not inbuf:`
```
if not inbuf:
            raise OrchestratorValidationError(
                'missing credential configuration file. Retry with -i <filename>')
```
and
```
if not inbuf and not all_available_devices:
            # one parameter must be present
            return HandleCommandResult(-errno.EINVAL, '--all-available-devices is required')
```

to me, they both seem to be handling cases where `-i` is not passed. In the first case, which is for SNMP, it's basically saying you MUST use `-i` so if there is no inbuf, there is a problem. The second case, which is for OSDs, is essentially requiring that you have ot pass either `-i` or `--all-available-devices`. We don't allow the command without exactly one of those being passed (it also has a check for `if inbuf and all_available_devices` before that). 

What I did also see was

```
if inbuf:
            raise OrchestratorValidationError('unrecognized command -i; -h or --help for usage')
```

Which all seems to me seem to imply `if not inbuf` means `-i` was not passed, and `if inbuf` implies `-i` was passed.

[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/orchestrator/module.py:1844]: OK, I guess I misremembered. Looking at the code I'm left totally confused, so I'm just going to try removing inbuf from the args and seeing what happens.

File: qa/suites/orch/cephadm/smb/tasks/deploy_smb_basic.yaml (hunk 16 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply qa/suites/orch/cephadm/smb/tasks/deploy_smb_basic.yaml]: fixed

**Comment by adk3798** [comment]
https://pulpito.ceph.com/phlogistonjohn-2024-03-22_03:02:08-orch:cephadm-wip-phlogistonjohn-testing-2024-03-21-1835-distro-default-smithi/

Vast majority is `XXX in cluster log` stuff, tests that weren't are all failures I've seen before (mds_upgrade_sequence, staggered upgrade with agent, etc.)

No reason to block merging this PR.

**Review by adk3798 — APPROVED** [review]
